cpp/src/replace/replace.cu

```diff
@@ -77,47 +102,97 @@ namespace{ //anonymous
   /* ----------------------------------------------------------------------------*/
   struct replace_kernel_forwarder {
     template <typename col_type>
-    void operator()(void*       d_col_data,
-                    gdf_size_type      nrows,
-                    const void* d_old_values,
-                    const void* d_new_values,
-                    gdf_size_type      nvalues)
+    void operator()(const gdf_column  &input_col,
+                    const gdf_column &old_values,
+                    const gdf_column &new_values,
+                    gdf_column       &output)
     {
-      thrust::device_ptr<const col_type> old_values_begin = thrust::device_pointer_cast(static_cast<const col_type*>(d_old_values));
+      const bool input_is_nullable = (input_col.valid != nullptr && input_col.null_count > 0);
+      const bool new_is_nullable = (new_values.valid != nullptr && new_values.null_count > 0);
+
+      const bit_mask_t* __restrict__ typed_input_valid = reinterpret_cast<bit_mask_t*>(input_col.valid);
+      const bit_mask_t* __restrict__ typed_new_valid = reinterpret_cast<bit_mask_t*>(new_values.valid);
+      bit_mask_t* __restrict__ typed_out_valid = reinterpret_cast<bit_mask_t*>(output.valid);
+
+      gdf_size_type *null_count = nullptr;
+      if (input_is_nullable || new_is_nullable) {
+          RMM_ALLOC(&null_count, sizeof(gdf_size_type), 0);
+          CUDA_TRY(cudaMemset(null_count, 0, sizeof(gdf_size_type)));
+       }
+
+      thrust::device_ptr<const col_type> old_values_begin = thrust::device_pointer_cast(static_cast<const col_type*>(old_values.data));
 
-      const size_t grid_size = nrows / BLOCK_SIZE + (nrows % BLOCK_SIZE != 0);
-      replace_kernel<<<grid_size, BLOCK_SIZE>>>(static_cast<col_type*>(d_col_data),
-                                             nrows,
+      cudf::util::cuda::grid_config_1d grid{input_col.size, BLOCK_SIZE, 1};
+
+      auto replace = replace_kernel<col_type, true, true>;
+
+      if(true == input_is_nullable && false == new_is_nullable){
```

Comment:
Since you picked such nice readable names as `input_is_nullable` and `new_is_nullable`, I think it'd actually be clearer to do...

```
if(input_is_nullable){
   if(new_is_nullable){
       replace = replace_kernel<col_type, true, true>;
   } else{
       replace = replace_kernel<col_type, true, false>;
   }
}else{
   if(new_is_nullable){
       replace = replace_kernel<col_type, false, true>;
   } else{
       replace = replace_kernel<col_type, false, false>;
   }
}

Reply:
And I'd rename `input_is_nullable` and `new_is_nullable` to `input_has_nulls` and `new_has_nulls`. Because you're not testing for whether or not a value _can_ be null (i.e., `col->valid` is allocated), but instead if any nulls actually are present.

Reply:
"is_nullable" is my fault. After discussion based on one of Eyal's PRs, it became clear that "nullable" is a common usage in arrow and other places...  "is_nullable" and "has_nulls" are not the same thing. "is_nullable" means the `column.valid != nullptr`. "has_nulls" (IMO) means `(column.valid != nullptr && column.null_count > 0)`